From 78eb59a6415395efeac0ce94939b1019421f6dd7 Mon Sep 17 00:00:00 2001
From: Tommy Huang <tommy_huang@aspeedtech.com>
Date: Tue, 9 Sep 2025 18:07:48 +0800
Subject: [PATCH 02/10] drivers: ast2700-mailbox: add waitting tx bit clear
 behavior

Add waitting tx trigger bit clear behavior in the
ast2700_mbox_send function. The tx timeout will be
assigned in the channel dtsi property.

Signed-off-by: Tommy Huang <tommy_huang@aspeedtech.com>
Change-Id: I821a9914a506c5f0e0b63ba0da7f6ec93a4066e3
(cherry picked from commit 4f2cc48cfea80b10eec7f3e8aa4ac304f14cb4b6)
Signed-off-by: Vasanthakumar K <vasanthakumark@ami.com>
---
 drivers/mailbox/ast2700-mailbox.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/mailbox/ast2700-mailbox.c b/drivers/mailbox/ast2700-mailbox.c
index 74277c56b1..042f80bb79 100644
--- a/drivers/mailbox/ast2700-mailbox.c
+++ b/drivers/mailbox/ast2700-mailbox.c
@@ -10,6 +10,7 @@
 #include <mailbox-uclass.h>
 #include <linux/bitfield.h>
 #include <linux/bug.h>
+#include <wait_bit.h>
 
 /* Each bit in the register represents an IPC ID */
 #define IPCR_TX_TRIG		0x00
@@ -28,6 +29,7 @@ struct ast2700_mbox_priv {
 	u8 msg_size;
 	void __iomem *tx_regs;
 	void __iomem *rx_regs;
+	u32 tx_tout;
 };
 
 static inline bool ast2700_mbox_tx_done(struct ast2700_mbox_priv *mb, int idx)
@@ -63,6 +65,10 @@ static int ast2700_mbox_send(struct mbox_chan *chan, const void *data)
 	writel(BIT(idx), mb->tx_regs + IPCR_TX_TRIG);
 	debug("%s: Ch-%d sent\n", __func__, idx);
 
+	/* Wait until observation bit is cleared */
+	wait_for_bit_le32((const void *)(mb->tx_regs + IPCR_TX_TRIG), BIT(idx), false,
+			  mb->tx_tout, false);
+
 	return 0;
 }
 
@@ -132,6 +138,8 @@ static int ast2700_mbox_probe(struct udevice *dev)
 		return -EINVAL;
 	priv->msg_size = data->msg_size;
 
+	priv->tx_tout = dev_read_u32_default(dev, "aspeed,tx-timeout", -1);
+
 	return 0;
 }
 
-- 
2.34.1

